static/src/javascripts/projects/common/modules/identity/account-profile.js

```diff
@@ -0,0 +1,262 @@
+            }
+        }
+    }
+
+    /*
+     *   Handle user avatar upload
+     *   TO DO: Use html5 file api to validate file size prior to upload @chrisfinch
+     */
+    bindAvatarUpload() {
+        const avatarForm = this.accountProfileForms.querySelector(
+            classes.avatarUploadForm
+        );
+
+        if (avatarForm) {
+            bean.on(avatarForm, 'submit', event => {
+                event.preventDefault();
+
+                avatarUploadByApi(((avatarForm: any): HTMLFormElement));
+            });
+        }
+    }
+
+    /*
+     *   Generate a descriptive error message for when a user attempts to leave
+     *   a form with unsaved changes.
+     */
+    genUnsavedError() {
+        let labelId;
+        let text;
+        const errorDivStart = '<div class="form__error">';
+        const errorDivEnd = '</div>';
+        const errorSaveLink =
+            '<a href="#" class="js-save-unsaved">Save changes</a>';
+        let errorMessageStart = 'Your form has unsaved changes in ';
+
+        for (let i = 0; i < this.unsavedFields.length; i += 1) {
+            labelId = this.unsavedFields[i].id;
+            const label = this.accountProfileForms.querySelector(
+                `[for="${labelId}"]`
+            );
+            if (label) {
+                text = label.innerHTML;
+                errorMessageStart += `"${text}"`;
+                if (i === this.unsavedFields.length - 1) {
+                    errorMessageStart += '. ';
+                } else if (i === this.unsavedFields.length - 2) {
+                    errorMessageStart += ' and ';
+                } else {
+                    errorMessageStart += ', ';
+                }
+            }
+        }
+
+        return errorDivStart + errorMessageStart + errorSaveLink + errorDivEnd;
+    }
+
+    /*
+     *   Register a form and form field as containing unsaved changes
+     */
+    onInputChange(event: Event) {
+        const input = ((event.target: any):
+            | HTMLInputElement
+            | HTMLTextAreaElement);
+        bonzo(input.form).addClass(classes.changed);
+        this.unsavedChangesForm = input.form;
+        if (!this.unsavedFields.some(el => el === input)) {
+            this.unsavedFields.push(input);
+        }
+    }
+
+    /*
+     *   Bind keyup events on input fields and register parent form on element
+     */
+    bindInputs(form: ?HTMLElement) {
+        if (form instanceof HTMLFormElement) {
+            const inputs = Array.prototype.slice.call(
+                form.querySelectorAll(classes.textInput)
+            );
+            inputs
+                .concat(
+                    Array.prototype.slice.call(form.querySelectorAll('select'))
+                )
+                .forEach(input => {
+                    if (input.type === 'select-one') {
+                        input.addEventListener(
+                            'change',
+                            this.onInputChange.bind(this)
```

Comment:
I believe this could be changed to

```js
input.addEventListener('change', event => this.onInputChange(event))
```

because of the fat brackets. Same for line 256.

Reply:
Is that generally preferred? Using `bind` at least has the virtue of being clear about what you're attempting to do, whereas otherwise you need to have some fairly clear idea of the fat arrow scoping rules are to understand why you can't just use `this.onInputChange`.

Reply:
I'm not sure, if it is generally preferred, but we tried to eliminate this pattern whereever it's possible for two reasons:

- [performance](https://jsperf.com/function-bind-performance/15)
- readability: I feel it's still harder to read than fat arrows, because with `.bind()` you also have to think about what `this` actually is in the context of the call.

Generally we'd prefer a solution without `this` or OOP - just passing around objects.

Reply:
Thanks @gustavpursche, will change. I agree about moving away from `this`, but not wanting to do too much in one PR.